### PR TITLE
Use forwarding

### PR DESCRIPTION
**Pull request by akva2**
This use class forwarding in more spots where possible.

Some of the commits are possibly a little more controversial. I have opted to use private template definitions, and possibly explicit template instantations, to avoid symbol usage in headers. I also had to duplicate some parsing code in the SICD/AutoICD.

https://github.com/OPM/opm-material/pull/455, https://github.com/OPM/opm-grid/pull/526, https://github.com/OPM/opm-simulators/pull/3186 and https://github.com/OPM/opm-models/pull/651 must be merged before this.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this opm-material=455 opm-models=651 please

**Comment by akva2** [comment]
jenkins build this opm-grid=526 opm-models=651 please

**Comment by joakim-hove** [comment]
jenkins build this opm-grid=526 please

**Comment by akva2** [comment]
jenkins build this opm-grid=526 opm-simulators=3186 please

File: opm/parser/eclipse/EclipseState/Schedule/MSW/SICD.hpp [hunk 1]
```diff
@@ -91,33 +92,6 @@ namespace Opm {
             serializer(m_scaling_factor);
         }
 
-        template<class ICD>
-        static std::map<std::string, std::vector<std::pair<int, ICD> > >
-        fromWSEG(const DeckKeyword& wseg) {
```
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/MSW/SICD.hpp:96]: This could have been a free function in a private header to avoid the code duplication?

Comment (akva2) [reply opm/parser/eclipse/EclipseState/Schedule/MSW/SICD.hpp:96]: ah yes, i somehow got the idea it accessed members. will improve.

Comment (akva2) [reply opm/parser/eclipse/EclipseState/Schedule/MSW/SICD.hpp:96]: done.

**Comment by akva2** [comment]
jenkins build this please

File: opm/parser/eclipse/EclipseState/Schedule/MSW/SICD.hpp (hunk 1 above)
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Schedule/MSW/SICD.hpp:96]: Thanks for the update - will merge when green.

**Pull request merged**